### ISSUE TITLE
state param url encoding problem

### ISSUE THREAD

**Issue by Windvis** [BitySA/oauth2-auth-code-pkce#2]
It looks like some OAuth Saas providers url encode characters in the state query param which shouldn't be encoded. For example, it encodes `~` as `&7E`.

Because of that url encoding, [this check](https://github.com/BitySA/oauth2-auth-code-pkce/blob/19b223652799a32a8f0252fd9578ff54490817a0/index.ts#L176) fails.

It's easily solvable by wrapping the [return value](https://github.com/BitySA/oauth2-auth-code-pkce/blob/19b223652799a32a8f0252fd9578ff54490817a0/index.ts#L453) in a `decodeURI` call, but I'm not sure if that's the best solution.

**Comment by lf94** [comment]
@Windvis could you please tell me which provider is doing this? :slightly_smiling_face:

**Comment by lf94** [comment]
If we take the RFC as the law (which it really should be!), then technically there should be no URI encoding. https://tools.ietf.org/html/rfc6749#section-4.2.2 See the part about the `state` parameter:

> REQUIRED if the "state" parameter was present in the client authorization request. **The exact value received from the client.**

On the other hand, I think `decodeURI` will not cause any damage to any state parameter generated, since it uses the URL-safe character set:

````ts
const PKCE_CHARSET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
```

I'll add the call in.

**Comment by lf94** [comment]
Actually doing more tests, your original statement seems to be incorrect.

```
>> encodeURIComponent('hello~mate')
< "hello~mate"
>> encodeURI('hello~mate')
< "hello~mate"
```

In both cases, the `~` was not encoded. So something else is going on.

**Comment by Windvis** [comment]
> @Windvis could you please tell me which provider is doing this? 🙂

Okta

> It also flew over my head that you said &7E and not %7E. I can't recall anything off the top of my head that uses the & encoding.

Sorry, I meant `%7E`, I miss typed when describing the issue.

> Actually doing more tests, your original statement seems to be incorrect. In both cases, the ~ was not encoded. So something else is going on. I won't continue further on this issue until we have some more information.

I'm guessing Okta url encodes on the back end and that implementation does the wrong thing? Anyway, you can verify it by creating an Okta dev account and generating a state value which includes a `~`.

**Comment by lf94** [comment]
@Windvis I'm overwhelmed by the amount of functionality presented to me in the Okta dashboard.

What is your setup for this service?

> I'm guessing Okta url encodes on the back end and that implementation does the wrong thing? 

Yes, I'm guessing too. Since the purpose of this package is to be universal, simple, but also RFC abiding, we will not change anything if that's the case. :stuck_out_tongue:

**Comment by Windvis** [comment]
> @Windvis I'm overwhelmed by the amount of functionality presented to me in the Okta dashboard.
> 
> What is your setup for this service?

I just followed [this guide](https://developer.okta.com/blog/2019/08/22/okta-authjs-pkce#create-an-oidc-application-on-okta) to set up the application.

> Yes, I'm guessing too. Since the purpose of this package is to be universal, simple, but also RFC abiding, we will not change anything if that's the case. 😛

Yea, that's totally fair! Maybe a simple warning in the readme will be enough? I'll try to find out how to contact the Okta devs themselves and get it fixed on their end.

**Comment by lf94** [comment]
> I just followed this guide to set up the application.

Excellent, thank you. I'm on it. :slightly_smiling_face:

**Comment by lf94** [comment]
@Windvis so here's the results.

1. Investigating the HTTP headers, it reveals they are using PHP.
2. Based on this, I made a guess they are using PHP's `urlencode()`.
3. ![1575395784](https://user-images.githubusercontent.com/547923/70076307-82c9db00-15cc-11ea-98dc-f6064579b3c0.png)
4. I checked to see if other languages are doing this, and yes, Python's `urllib.parse.urlencode()` does too.

I'll add it since it should really do no harm.
